### ISSUE TITLE
TypeError: state.popChainContext is not a function

### ISSUE THREAD

**Issue by aminya** [eslint/eslint#14078]
```js
> eslint .


Oops! Something went wrong! :(

ESLint: 7.19.0

TypeError: state.popChainContext is not a function
Occurred while linting C:\Users\aminy\atom-ide\nuclide\nuclide-commons\fsPromise.js:345
    at processCodePathToExit (C:\Users\aminy\atom-ide\node_modules\.pnpm\eslint@7.19.0\node_modules\eslint\lib\linter\code-path-analysis\code-path-analyzer.js:512:19)
    at CodePathAnalyzer.leaveNode (C:\Users\aminy\atom-ide\node_modules\.pnpm\eslint@7.19.0\node_modules\eslint\lib\linter\code-path-analysis\code-path-analyzer.js:729:9)
    at CodePathAnalyzer.overrideLeaveNode (C:\Users\aminy\atom-ide\node_modules\.pnpm\eslint-plugin-node@11.1.0_eslint@7.19.0\node_modules\eslint-plugin-node\lib\rules\process-exit-as-throw.js:130:27)
    at C:\Users\aminy\atom-ide\node_modules\.pnpm\eslint@7.19.0\node_modules\eslint\lib\linter\linter.js:954:32
    at Array.forEach (<anonymous>:null:null)
    at runRules (C:\Users\aminy\atom-ide\node_modules\.pnpm\eslint@7.19.0\node_modules\eslint\lib\linter\linter.js:947:15)
    at Linter._verifyWithoutProcessors (C:\Users\aminy\atom-ide\node_modules\.pnpm\eslint@7.19.0\node_modules\eslint\lib\linter\linter.js:1173:31)
    at Linter._verifyWithConfigArray (C:\Users\aminy\atom-ide\node_modules\.pnpm\eslint@7.19.0\node_modules\eslint\lib\linter\linter.js:1271:21)
    at Linter.verify (C:\Users\aminy\atom-ide\node_modules\.pnpm\eslint@7.19.0\node_modules\eslint\lib\linter\linter.js:1226:25)
    at LinterPrototype.verify (C:\Users\aminy\atom-ide\node_modules\.pnpm\eslint-plugin-only-warn@1.0.2\node_modules\eslint-plugin-only-warn\src\only-warn.js:14:55)
    at Linter.verifyAndFix (C:\Users\aminy\atom-ide\node_modules\.pnpm\eslint@7.19.0\node_modules\eslint\lib\linter\linter.js:1416:29)
    at verifyText (C:\Users\aminy\atom-ide\node_modules\.pnpm\eslint@7.19.0\node_modules\eslint\lib\cli-engine\cli-engine.js:239:48)
    at CLIEngine.executeOnFiles (C:\Users\aminy\atom-ide\node_modules\.pnpm\eslint@7.19.0\node_modules\eslint\lib\cli-engine\cli-engine.js:807:28)
    at ESLint.lintFiles (C:\Users\aminy\atom-ide\node_modules\.pnpm\eslint@7.19.0\node_modules\eslint\lib\eslint\eslint.js:521:23)
    at Object.execute (C:\Users\aminy\atom-ide\node_modules\.pnpm\eslint@7.19.0\node_modules\eslint\lib\cli.js:294:36)
    at main (C:\Users\aminy\atom-ide\node_modules\.pnpm\eslint@7.19.0\node_modules\eslint\bin\eslint.js:142:52)
    at Object.<anonymous> (C:\Users\aminy\atom-ide\node_modules\.pnpm\eslint@7.19.0\node_modules\eslint\bin\eslint.js:146:2)
    at Module._compile (node:internal/modules/cjs/loader:1108:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1137:10)
    at Module.load (node:internal/modules/cjs/loader:973:32)
    at Function.Module._load (node:internal/modules/cjs/loader:813:14)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:76:12)
    at node:internal/main/run_main_module:17:47

```

**Tell us about your environment**

* **ESLint Version:** 7.19.0
* **Node Version:** 15.8.0
* **npm Version:** 6.14.11

**What parser (default, `@babel/eslint-parser`, `@typescript-eslint/parser`, etc.) are you using?** 

I am using [eslint-config-atomic](https://github.com/atom-community/eslint-config-atomic/) which uses [babel-eslint](https://github.com/atom-community/eslint-config-atomic/blob/ed22dc8ed00a81dd0b71deb123bac0b4ccb95d3e/index.js#L41)

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```json
{
  "extends": "eslint-config-atomic",
  "ignorePatterns": ["dist/", "node_modules/"]
}

```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

https://github.com/atom-community/atom-ide-base/blob/8e952b465a6485eee76881cc201108e5bd853668/nuclide/nuclide-commons/fsPromise.js#L345
```js
const mode = typeof options !== 'string' ? options?.mode : null;
```

```bash
eslint .
```

**What did you expect to happen?**
Lint

**What actually happened? Please include the actual, raw output from ESLint.**
An internal Eslint error happens

**Are you willing to submit a pull request to fix this bug?**
Not sure how to fix it

Mirror: https://github.com/atom-community/atom-ide-base/issues/56

**Comment by mdjermanovic** [comment]
Hi @aminya, thanks for the issue!

I can reproduce this. It seems to be caused by `eslint-plugin-coffee` modifying some internal prototypes [here](https://github.com/helixbass/eslint-plugin-coffee/blob/master/src/patch-code-path-analysis.coffee) (probably for the purpose of linting CoffeeScript source files, but it apparently affects everything).

The error doesn't happen when I remove [override](https://github.com/atom-community/eslint-config-atomic/blob/8e3d7e9349be1c5572f933fb051a367685375ff3/index.js#L98) that uses `eslint-plugin-coffee` and run `eslint .`, or if I run eslint directly on `nuclide/nuclide-commons/fsPromise.js`.

**Comment by aminya** [comment]
Tagging the author of `eslint-plugin-coffee` (@helixbass), so they know about this.

**Comment by helixbass** [comment]
Ah thanks for the head start on this @mdjermanovic @aminya !

That's obviously problematic if that monkeypatching from `eslint-plugin-coffee` is causing problems elsewhere (fwiw the basic reason for the monkeypatching is that ESLint doesn't expose the ability to legitimately override its code path analysis behavior to custom parsers)

I'll start looking into this tonight

**Comment by helixbass** [comment]
Ok @aminya I took a stab at dynamically restricting the monkeypatching to only "kick in" when linting Coffeescript files, can you try installing this dev version of `eslint-plugin-coffee`:
```
yarn add --dev github:helixbass/eslint-plugin-coffee#eslint-plugin-coffee-v0.1.14-dev.1-gitpkg
```
And let me know if that resolves the issue? If it does, I can bump that version to beta and then release

Also since this isn't an ESLint issue we can move discussion if that's preferable, eg I opened https://github.com/helixbass/eslint-plugin-coffee/issues/48

**Comment by aminya** [comment]
Thank you! Tested it, and the issue is fixed with that dev branch.

Yes, that is a better solution. We should not assume that everyone uses eslint-plugin-coffee only with CoffeeScript projects. Actually, [eslint-config-atomic](https://github.com/atom-community/eslint-config-atomic/) is configured to support as many as languages possible and not include any opinionated rule. 

If you are using the file extension to dynamically detect CoffeeScript, could you add `.cson` to the supported extensions too?

**Comment by helixbass** [comment]
Great, I'll mention when I release a new `eslint-plugin-coffee` version that includes that fix so that you can update dependencies

> If you are using the file extension to dynamically detect CoffeeScript, could you add `.cson` to the supported extensions too?

It's detecting CoffeeScript based on AST properties set by `eslint-plugin-coffee`, so it should also pick up `.cson` since it appears `eslint-config-atomic` is already mapping that extension to `eslint-plugin-coffee`

**Comment by helixbass** [comment]
Released `eslint-plugin-coffee@0.1.14` which contains the above fix

So I think this can be closed?
